db/event_manager_test.go

```diff
@@ -831,3 +831,126 @@ func assertChannelLengthWithTimeout(t *testing.T, c chan interface{}, expectedLe
+	url := ts.URL
+	wr.Clear()
+	em := NewEventManager()
+	em.Start(5, 100)
+
+	source := `function(doc) { if (doc.state == "offline") { return true; } else { return false; } }`
+	timeout := uint64(30)
+	wh, _ := NewWebhook(fmt.Sprintf("%s/echo", url), source, &timeout)
+	em.RegisterEventHandler(wh, DBStateChange)
+
+	warnCountPrev, err := strconv.Atoi(base.StatsResourceUtilization().Get(base.StatKeyWarnCount).String())
+	assert.NoError(t, err, "No error while getting the count of warnings")
+
+	for i := 0; i < 20; i += 2 {
+		err := em.RaiseDBStateChangeEvent("db", "online", "Index service is listening", "127.0.0.1:4985")
+		assert.NoError(t, err, "Database state change event should be raised successfully")
+		err = em.RaiseDBStateChangeEvent("db", "offline", "Index service heartbeat lost!", "127.0.0.1:4985")
+		assert.NoError(t, err, "Database state change event should be raised successfully")
+	}
+	func() {
+		timeout := time.After(1000 * time.Millisecond)
+		tick := time.Tick(100 * time.Millisecond)
+		for {
+			select {
+			case <-timeout:
+				log.Fatalf("Retry timedout while reading request count and payload!")
+				return
+			case <-tick:
+				if wr.GetCount() == 10 && len(wr.GetPayloads()) == 10 {
+					for i, s := range wr.GetPayloads() {
+						fmt.Printf("Posted payload-%v:%v\n", i, string(s))
+						assert.NotEmpty(t, string(s), "Request body should not be empty")
+						assert.Contains(t, string(s), `"state":"offline"`)
+						assert.NotContains(t, string(s), `"state":"online"`)
+					}
+					return
+				}
+			}
+		}
+	}()
+	warnCountCurr, err := strconv.Atoi(base.StatsResourceUtilization().Get(base.StatKeyWarnCount).String())
+	assert.NoError(t, err, "No error while getting the count of warnings")
+	warnCount := warnCountCurr - warnCountPrev
+	assert.Equal(t, 10, warnCount, "10 online events should be skipped and 10 offline events should be posted")
+}
+
+// Simulate the scenario for filter function error. Create 10 database status change (online) events with bad
+// syntax in filter function. Filter function processing should be aborted and no events should be posted.
+func TestWebhookHandleEventDBStateChangeFilterFuncError(t *testing.T) {
+	defer base.SetUpTestLogging(base.LevelInfo, base.KeyEvents)()
+	ts, wr := InitWebhookTest()
+	defer ts.Close()
+	url := ts.URL
+	wr.Clear()
+	em := NewEventManager()
+	em.Start(2, 100)
+
+	warnCountPrev, err := strconv.Atoi(base.StatsResourceUtilization().Get(base.StatKeyWarnCount).String())
+	assert.NoError(t, err, "No error while getting the count of warnings")
+
+	source := `func (doc) { if (doc.state == "offline") { return true; } else { return false; } }`
+	timeout := uint64(30)
+	wh, _ := NewWebhook(fmt.Sprintf("%s/echo", url), source, &timeout)
+	em.RegisterEventHandler(wh, DBStateChange)
+	err = em.RaiseDBStateChangeEvent("db", "offline", "Index service is listening", "127.0.0.1:4985")
+	assert.NoError(t, err, "Error calling filter function")
+
+	func(warnCountPrev int) {
```

Comment:
Pulling this into a helper function that can be used by multiple tests to block waiting for a target warnCount seems like a better approach.  (moving lines 921/922 out of the func and into this test).

Reply:
As discussed, this code becomes obsolete when we rewrite the init tests with the direct use of webhook handleEvent method.